src/pages/Defi/hooks/useCosmosStakingBalances.tsx

```diff
@@ -0,0 +1,187 @@
+import { CAIP19 } from '@shapeshiftoss/caip'
+import { chainAdapters, ChainTypes } from '@shapeshiftoss/types'
+import { useEffect, useMemo } from 'react'
+import { BigNumber, bnOrZero } from 'lib/bignumber/bignumber'
+import {
+  selectAccountSpecifier,
+  selectAssetByCAIP19,
+  selectMarketDataById,
+} from 'state/slices/selectors'
+import {
+  ActiveStakingOpportunity,
+  selectActiveStakingOpportunityDataByAssetId,
+  selectNonloadedValidators,
+  selectSingleValidator,
+  selectStakingDataIsLoaded,
+  selectValidatorIsLoaded,
+} from 'state/slices/stakingDataSlice/selectors'
+import { stakingDataApi } from 'state/slices/stakingDataSlice/stakingDataSlice'
+import { useAppDispatch, useAppSelector } from 'state/store'
+
+const SHAPESHIFT_VALIDATOR_ADDRESS = 'cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf'
+
+type UseCosmosStakingBalancesProps = {
+  assetId: CAIP19
+}
+
+export type UseCosmosStakingBalancesReturn = {
+  activeStakingOpportunities: MergedActiveStakingOpportunity[]
+  stakingOpportunities: MergedStakingOpportunity[]
+  totalBalance: string
+  isLoaded: boolean
+}
+
+export type MergedActiveStakingOpportunity = ActiveStakingOpportunity & {
+  fiatAmount?: string
+  tokenAddress: string
+  assetId: CAIP19
+  chain: ChainTypes
+  tvl: string
+}
+
+export type MergedStakingOpportunity = chainAdapters.cosmos.Validator & {
+  tokenAddress: string
+  assetId: CAIP19
+  chain: ChainTypes
+  tvl: string
+}
+
+export function useCosmosStakingBalances({
+  assetId,
+}: UseCosmosStakingBalancesProps): UseCosmosStakingBalancesReturn {
+  const isStakingDataLoaded = useAppSelector(selectStakingDataIsLoaded)
+  const isValidatorDataLoaded = useAppSelector(selectValidatorIsLoaded)
+  const isLoaded = isStakingDataLoaded && isValidatorDataLoaded
+  const marketData = useAppSelector(state => selectMarketDataById(state, assetId))
+  const asset = useAppSelector(state => selectAssetByCAIP19(state, assetId))
+  const dispatch = useAppDispatch()
+
+  const accountSpecifiers = useAppSelector(state => selectAccountSpecifier(state, asset?.caip2))
+  const accountSpecifier = accountSpecifiers?.[0]
+
+  const activeStakingOpportunities = useAppSelector(state =>
+    selectActiveStakingOpportunityDataByAssetId(
+      state,
+      accountSpecifier,
+      SHAPESHIFT_VALIDATOR_ADDRESS,
+      asset.caip19,
+    ),
+  )
+
+  const shapeshiftValidator = useAppSelector(state =>
+    selectSingleValidator(state, accountSpecifier, SHAPESHIFT_VALIDATOR_ADDRESS),
+  )
+  const stakingOpportunities = useMemo(() => {
+    return [
+      {
+        ...shapeshiftValidator,
+      },
+    ]
+  }, [shapeshiftValidator])
+
+  const nonLoadedValidators = useAppSelector(state =>
+    selectNonloadedValidators(state, accountSpecifier),
+  )
+
+  const chainId = asset.caip2
+
+  const mergedActiveStakingOpportunities = useMemo(() => {
```

Comment:
I'd love to see tests for all the functions that do math, at least.

Reply:
Added tests covering this file !

Reply:
Tests looking good and they're testing in the context of react, but in the future, it would be nice to:
1. Make the content of this cope `useMemo () => {}` its own function so it can be tested as a unit (doesn't have to be done in this PR, the way you're testing is also testing the same thing the "react way"
2. Switch to `useCallback` since that's what it effectively is